### ISSUE TITLE
Order of nested blocks seems wrong

### ISSUE THREAD

**Issue by SamyPesse** [mitermayer/draft-js#30]
I've updated the [build](https://github.com/SamyPesse/draft-js/tree/build) branch which is a merge of the `feature/tree` branch and the metadata PR.

Since the last update, the order of nested blocks rendered seems wrong:

Output:

<img width="857" alt="screen shot 2016-06-18 at 23 25 40" src="https://cloud.githubusercontent.com/assets/845425/16173765/50aa7bec-35ac-11e6-9170-58832d8585f7.png">

ContentState dump:

``` js
       "00041": {
            "key": "00041",
            "type": "table",
            "text": "",
            "characterList": [],
            "depth": 0,
            "data": {}
        },
        "00041/00036": {
            "key": "00041/00036",
            "type": "table-header",
            "text": "",
            "characterList": [],
            "depth": 0,
            "data": {}
        },
        "00041/00036/00035": {
            "key": "00041/00036/00035",
            "type": "table-row",
            "text": "",
            "characterList": [],
            "depth": 0,
            "data": {}
        },
        "00041/00036/00035/00033": {
            "key": "00041/00036/00035/00033",
            "type": "table-cell",
            "text": " Online book about how to write a computer operating system in C/C++ from scratch. ",
```

The first cell is the one with the long text in the `blockMap`, but order is wrong in the rendering.

@mitermayer Is it related to the perfs improvements ? How can I help ?

**Comment by mitermayer** [comment]
@SamyPesse will have a look into this, can you provide me the data that was used to initialize the editor ?

**Comment by mitermayer** [comment]
The performance improvements goal was to have one function that iterates the minimum amount of time on the blockMap and creates something like this

```
{
   $key: {
      firstLevelBlock: Immutable.OrderedMap({ key: ContentBlock} ...),
      childrenBlocks: Immutable.Set([ContentBlock...])
  }
}
```

The goal is being able to use that for checks on `shouldComponentUpdate()` of the blocks so that it only needs to render if some childrens has changed by comparing `childrenBlocks`. And also by avoiding to `getFirstLevelBlocks()` and `getBlockChildren()` by using the `BlockMapTree.getIn([key, 'firstLevelBlocks'])`. The idea is that the `BlockMapTree` should be generated/updated only once per render, and passed across as props to inner children to avoid iteration on the blockMap.

**Comment by mitermayer** [comment]
The algorithm is on `ContentState.getBlockDescendants()`

It has been done on reverse order so that we can avoid backTracking to include children for already computed blocks.  The gotcha is that on the end we need to reverse the blockMap for the block since it has been added backwards.

Still playing around with this algorithmn and data structure. Before it we were 10x slower than master even without nesting blocks, now we are 1.9x slower. Will try to reduce as much as possible performance impact.

During this work I also identified potential enhancements for draft master performance algorithms that i will create a separate PR for and cherry pick in here.

Please let me know what are your thoughts on it. Feel free to jump in and contribute with this issue if you can. I will have a bit less time to keep working on this PR since i will only be able to do during my night times

**Comment by mitermayer** [comment]
@SamyPesse  could you try this again with my latest commit ?

**Comment by mitermayer** [comment]
same as https://github.com/SamyPesse/draft-js/issues/29 please check again
